IC-16 : Add JWT Authentication to API

sobuse opened the pull request:
* Installed Microsoft.AspNetCore.Identity.EntityFrameworkCore nuget package
* Refactored my DBContext class to Inherit from IdentityDbContext<User>
* Added ConfigureIdentity method to my services method

sobuse committed:
Inherit Identity User

sobuse committed:
downloaded identity package

sobuse committed:
configured Identity

sobuse commented:
When I ran migration I got this error

![image](https://user-images.githubusercontent.com/59874300/184214672-8ea143e3-846f-4c22-8cb4-50f91516311f.png)

sobuse commented:
I tried fixing it with this [solution]( https://stackoverflow.com/questions/59796411/unable-to-create-an-object-of-type-applicationdbcontext-for-the-different-pat) it did not work

ehimah commented:
> When I ran migration I got this error
> 
> ![image](https://user-images.githubusercontent.com/59874300/184214672-8ea143e3-846f-4c22-8cb4-50f91516311f.png)

you missed step 5 of the task. read the book pg 245 and 246 again... and you'll see that you were supposed to create a method in your now startup class


```cs
public static void ConfigureIdentity(this IServiceCollection services) {
    var builder = services.AddIdentityCore<User>(o => {
         o.Password.RequireDigit = true;
          o.Password.RequireLowercase = false;
           o.Password.RequireUppercase = false;
            o.Password.RequireNonAlphanumeric = false;
             o.Password.RequiredLength = 10;
              o.User.RequireUniqueEmail = true;
        }); 
    builder = new IdentityBuilder(builder.UserType, typeof(IdentityRole), builder.Services); 
    builder.AddEntityFrameworkStores<RepositoryContext>() 
        .AddDefaultTokenProviders(); 
}
```

sobuse commented:
oooh

sobuse committed:
Reconfigured my startup class

sobuse committed:
Update InstagramCloneDbContext.cs

sobuse commented:
I have done the needful , After running migration I got a different migration error

![image](https://user-images.githubusercontent.com/59874300/184352837-0ca36740-882e-494c-b80e-ef7c505ec541.png)

ehimah commented:
The thing is that you need to change the IdentityUser class primary key to Guid.

follow the advice in the articles I sent to you. I just re-pasted them below

https://bel62.wordpress.com/2021/07/12/change-identity-to-use-guid-in-in-ef-core-3-1/

https://stackoverflow.com/a/61908959

sobuse commented:
Ok

sobuse committed:
Created application Roles class

sobuse committed:
created application User class

sobuse committed:
updated dbcontext class

sobuse committed:
Updated my startup class

sobuse committed:
Update UsersController.cs

sobuse commented:
I had this error after I did the configuration

![image](https://user-images.githubusercontent.com/59874300/184986266-06964b6b-c1f7-4888-b4a0-bc2fdec91c3e.png)

ehimah commented:
> I had this error after I did the configuration
> 
> ![image](https://user-images.githubusercontent.com/59874300/184986266-06964b6b-c1f7-4888-b4a0-bc2fdec91c3e.png)

yes, you have to re-do your user seeding configuration class. Please read this article for ideas on how to model the user and the password hasher

https://dnilvincent.com/blog/posts/seed-users-and-roles-in-asp-net-core-with-entity-framework-core

sobuse committed:
Added Password Hasher

sobuse committed:
reconfigured context class

sobuse committed:
deleted application user class

sobuse committed:
removed white space and inherit Identity user with guid

sobuse committed:
Reconfigured startup class

sobuse committed:
cleared up code

sobuse commented:
After  reconfiguring  my code then I tried to run the migration  , I had this error
![image](https://user-images.githubusercontent.com/59874300/185256492-8769b710-b2ff-4fb7-9f52-198eab68a031.png)

sobuse committed:
Changed to Identity Password

sobuse committed:
Update User.cs

sobuse committed:
deleted migration folder

sobuse committed:
added extention class for roles configuration

sobuse committed:
called my seed roles method

sobuse committed:
changed from dbContext to identityDbContext

sobuse committed:
added migration

sobuse committed:
changed position where i called configuredIdentity method

sobuse committed:
Update ApplicationRoleConfiguration.cs

sobuse committed:
changed password

sobuse committed:
added migration

sobuse committed:
Update InstagramCloneDbContext.cs

sobuse committed:
reconfigured Identity

sobuse committed:
Change my db name

sobuse committed:
Configured identity roles

sobuse committed:
reconfigured my user to use identity properties

sobuse committed:
dropped my database so I had to also delete migration

sobuse committed:
Added new migration to get identity tables on my db

sobuse committed:
Update InstagramCloneDbContext.cs

sobuse committed:
used identity user

sobuse committed:
Update InstagramCloneDbContextModelSnapshot.cs

sobuse committed:
reconfigured my add identity method

sobuse committed:
added new migration

sobuse committed:
changed the name userManager to User

sobuse committed:
reconfigured create user method in user controller class

sobuse committed:
Update UserCreateDTO.cs

sobuse committed:
changed position where i called my addController method

sobuse committed:
configured jwt in appsetting

sobuse committed:
created authentication manager and added authentication method

sobuse committed:
created jwt extension method

sobuse committed:
Create AuthenticationController.cs

sobuse committed:
created authentication Dto class

sobuse committed:
installed jwt bearer package

sobuse committed:
called my jwt extension method in the startup class

sobuse committed:
Update AuthenticationManager.cs

sobuse committed:
Update AuthenticationController.cs

sobuse committed:
injected authentication manager

sobuse committed:
Added authorization and formatted code

sobuse committed:
reconfigured code base

sobuse committed:
reconfigured code base

sobuse committed:
cleaned up code base and added allowAnonymous to my login controller

sobuse committed:
Update PostController.cs

Used asyn and await on the controller method

changed the authorId to a login authorId

sobuse committed:
Update appsettings.json

ehimah commented on src/InstagramClone.Api/Database/InstagramCloneDbContext.cs:
```diff
@@ -1,12 +1,13 @@
 namespace InstagramClone.Api.Database
 {
-    public class InstagramCloneDbContext : DbContext
+    public class InstagramCloneDbContext : IdentityDbContext<User, ApplicationRole, Guid>//IdentityDbContext<User> // IdentityDbContext<ApplicationUser> //DbContext
     {
```
clean up left over comment

sobuse replied:
Done

ehimah commented on src/InstagramClone.Api/Database/InstagramCloneDbContext.cs:
```diff
@@ -15,9 +16,10 @@ public InstagramCloneDbContext(DbContextOptions options) : base(options)
             
 
-            modelBuilder.Entity<User>().HasKey(u => u.Id);
+           // modelBuilder.Entity<User>().HasKey(u => u.Id);
             modelBuilder.Entity<Post>().HasKey(p => p.Id);
```
clean up and format this file

sobuse replied:
Done

ehimah commented on src/InstagramClone.Api/Database/InstagramCloneDbContext.cs:
```diff
@@ -61,17 +63,18 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
         
        
-        public DbSet<User> Users { get; set; }
+       //public DbSet<User> Users { get; set; }
```
clean up

sobuse replied:
Done

ehimah commented on src/InstagramClone.Api/Entities/User.cs:
```diff
@@ -1,19 +1,19 @@
-        
-        public Guid Id { get; set; } //PK
-        public string Email { get; set; }
-        public string Password { get; set; }
+       // public Guid Id { get; set; } //PK
```
clean up and format

sobuse replied:
Done

ehimah commented on src/InstagramClone.Api/Startup.cs:
```diff
@@ -37,20 +50,76 @@ private static void ConfigureService(WebApplicationBuilder builder)
             app.MapGet("/", () => "Hello World!");
 
         }
+
+        //public static void JwtConfiguration(this IServiceCollection services, IConfiguration configuration)
+        //{
+        //    var jwtSettings = configuration.GetSection("jwtSettings");
+        //    var secretKey = Environment.GetEnvironmentVariable("SECRET");
+
+        //    services.AddAuthentication(opt =>
+        //    {
+        //        opt.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
```
clean up and format this file

sobuse replied:
Done

ehimah reviewed (changes requested):
still, some changes left here. 

clean up and format the file

sobuse committed:
cleaned up code base

sobuse committed:
cleaned up code base and removed white spaces

sobuse committed:
cleaned up code base and removed white space

sobuse committed:
cleaned up and removed white space

sobuse commented on src/InstagramClone.Api/Controllers/UsersController.cs:
```diff
@@ -92,7 +118,7 @@ public IActionResult Follow([FromBody]FollowersDto followersDto)
             var isAlreadyFollowed = _context.UserFollowers.Any(uf => uf.FollowedUserId == followersDto.FollowedUserId
              && uf.FollowerId == followersDto.FollowerId);
```
This part is confusing when you say "FollowersDTO class should no longer contain a reference to the Follower / FollowerId" are you asking me to delete it from the dto class?

Or you meant this?
![image](https://user-images.githubusercontent.com/59874300/192361396-4fc53ef6-5d0c-4445-9893-f1e39f20680a.png)

ehimah replied:
You should delete it from the class and use the code sample I gave you. 

So we're saying that 

- the `followerId` will now be the logged-in user's id.
- We currently take that value from the HTTP Request DTO, but you should change the logic to use the id of the logged-in user.
- The `FollowedUserId` (the person the logged-in user wants to follow) will remain in the DTO.

sobuse replied:
Done.

sobuse commented on src/InstagramClone.Api/Controllers/UsersController.cs:
```diff
@@ -36,35 +44,53 @@ public IActionResult GetUser(Guid id)
         [HttpPost("")]
-        public IActionResult CreateUsers([FromBody] UserCreateDTO userDto)
+        [AllowAnonymous]
+        //[ServiceFilter(typeof(ValidationFilterAttribute))]
+        public async Task<IActionResult> RegisterUser([FromBody] UserCreateDTO userDto)
```
Done.

sobuse commented on src/InstagramClone.Api/Controllers/UsersController.cs:
```diff
@@ -2,27 +2,35 @@
         {
             // user variable is used to store the value of a particuler user from the database
             // This finds a user by id from the db set user property from the context class
+            // var user = userManager.GetUserId(id)
             var user = _context.Users.Find(id);
             if (user == null)
             {
```
Done

sobuse commented on src/InstagramClone.Api/Controllers/PostController.cs:
```diff
@@ -29,34 +34,34 @@ public IActionResult PostedItem(Guid id)
                 Content = postCreateDto.Content,
-                AuthorId = postCreateDto.AuthorId,
+                AuthorId = loggedInUser.Id,
+               // AuthorId = postCreateDto.AuthorId,
                 CreatedDate = postCreateDto.CreatedDate,
```
Done

sobuse committed:
cleaned up code base

sobuse committed:
refactored code base to accept logged in  user

sobuse committed:
refactored followersDto class

sobuse committed:
Update appsettings.json

sobuse commented on src/InstagramClone.Api/Configuration/UserConfiguration.cs:
```diff
@@ -8,16 +9,19 @@ public static class UserConfiguration
                         FirstName = "Ehueriken",
                         LastName = "Obuse",
-                        Avatar = "Sammy.jpeg",
+                        Avatar = "Sammy.jpeg",                    
```
Ok

sobuse committed:
changed picture from server storage to fetching from cloud

ehimah commented on src/InstagramClone.Api/DTOs/UserForAuthenticationDto.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public class UserForAuthenticationDto
+    {
+        [Required(ErrorMessage="Email is required")]
+        public string Email { get; set; }
+
+        [Required(ErrorMessage = "Password is required")]
+        public string Password { get; set; }
```
👌

ehimah commented on src/InstagramClone.Api/Entities/User.cs:
```diff
@@ -1,19 +1,17 @@
         public ICollection<UserFollower> Followed { get; set; } // Navigation
         public ICollection<PostReaction> Reactions { get; set; } // Navigation
-
+                                                                 // 
```
clean up

sobuse replied:
Done.

ehimah commented on src/InstagramClone.Api/appsettings.json:
```diff
@@ -6,7 +6,12 @@
   },
   "ConnectionStrings": {
-    "DefaultConnectionString": "Server=.;Database=InstagramClone2;Trusted_Connection=True;MultipleActiveResultSets=true"
+    "DefaultConnectionString": "Server=.;Database=InstagramClone3;Trusted_Connection=True;MultipleActiveResultSets=true"
```
you should drop all these old database versions and re-run your migration so that this stay as `InstagramClone`

sobuse replied:
Done.

ehimah commented on src/InstagramClone.Api/Configuration/ConfigurationOfJwt.cs:
```diff
@@ -0,0 +1,37 @@
+        {
+            var jwtSettings = configuration.GetSection("JwtSettings");
+            var secretKey = Environment.GetEnvironmentVariable("SECRET");
+            //var secretKey = "SECRET";
+
```
clean up

sobuse replied:
Done

ehimah commented on src/InstagramClone.Api/Authentication/AuthenticationManager.cs:
```diff
@@ -0,0 +1,86 @@
+                    issuer: jwtSettings.GetSection("validIssuer").Value,
+                    audience: jwtSettings.GetSection("validAudience").Value,
+                    claims: claims,
+                    expires: DateTime.Now.AddMinutes(Convert.ToDouble(jwtSettings.GetSection("expires").Value)),
+                    signingCredentials: signingCreadentails
```
you can change this to hours `DateTime.Now.AddHours(...` to give you more token time

sobuse replied:
Done.

ehimah replied:
you should comment "Done" only after you have pushed your commits

sobuse replied:
Ok. I comment the done on them to enable me to track the ones I have implemented because they are quite a few and  I would not want to miss any comments. I will look for a better way to track them in the next PR. Thanks a lot.

ehimah reviewed (commented):
This is looking good!

please action these final review comments and run the formatted on all the files for consistency

sobuse commented on src/InstagramClone.Api/DTOs/UserCreateDTO.cs:
```diff
@@ -1,13 +1,19 @@
         public Guid Id { get; set; } //PK
         public string Email { get; set; }
-        public string Password { get; set; }
+        [Required (ErrorMessage ="Password is required.")]
+        public string Password { get; set; } 
```
Ok.

sobuse committed:
changed db name

sobuse committed:
changed jwt expiration time

sobuse committed:
cleaned up code base

sobuse committed:
Added a new migration

sobuse committed:
Added validations

ehimah reviewed (approved):
🥳🥳🥳
